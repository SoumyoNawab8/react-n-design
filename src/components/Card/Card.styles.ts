import styled, { css, keyframes } from 'styled-components';

// --- Re-usable Spinner ---
const spin = keyframes`to { transform: rotate(360deg); }`;
export const Spinner = styled.div`
  border: 3px solid rgba(0, 0, 0, 0.2);
  /* 2. Access theme from props */
  border-top-color: ${({ theme }) => theme.colors.primary};
  border-radius: 50%;
  width: 2.5em;
  height: 2.5em;
  animation: ${spin} 1s linear infinite;
`;

export const LoadingOverlay = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  /* 3. Use theme color for background with alpha transparency */
  background: ${({ theme }) => theme.colors.background}b3; /* b3 is ~70% opacity */
  display: flex;
  justify-content: center;
  align-items: center;
  z-index: 10;
  border-radius: ${({ theme }) => theme.borderRadius};
`;
// --- End Spinner ---

const paddings = {
  none: '0px',
  small: '16px',
  medium: '24px',
  large: '32px',
};

export const CardCover = styled.div`
  width: 100%;
  img {
    display: block;
    width: 100%;
  }
`;

export const CardHeader = styled.div`
  margin-bottom: 16px;
  font-size: 1.25rem;
  font-weight: 600;
`;

export const CardBody = styled.div<{ padding: 'none' | 'small' | 'medium' | 'large' }>`
  padding: ${({ padding }) => paddings[padding]};
`;

export const CardFooter = styled.div`
  margin-top: 16px;
  /* 4. Use a theme color for borders */
  border-top: 1px solid ${({ theme }) => theme.colors.shadowDark}40; /* 40 is ~25% opacity */
  padding-top: 16px;
`;

export const StyledCard = styled.div<{
  variant: 'outset' | 'inset';
  bordered: boolean;
  hoverable: boolean;
  isLoading: boolean;
}>`
  position: relative;
  /* 5. Access all theme values from props */
  background: ${({ theme }) => theme.colors.background};
  border-radius: ${({ theme }) => theme.borderRadius};
  color: ${({ theme }) => theme.colors.text};
  box-shadow: ${({ variant, theme }) =>
    variant === 'inset' ? theme.shadows.softInset : theme.shadows.soft};
  overflow: hidden;
  transition: all 0.2s ease-in-out;

  ${({ bordered, theme }) =>
    bordered &&
    css`
      border: 1px solid ${theme.colors.shadowDark}40;
    `}

  ${({ hoverable }) =>
    hoverable &&
    css`
      cursor: pointer;
      &:hover {
        transform: translateY(-4px);
        /* 6. Make hover shadow theme-aware */
        box-shadow: 10px 10px 20px ${({ theme }) => theme.colors.shadowDark}, -10px -10px 20px ${({ theme }) => theme.colors.shadowLight};
      }
    `}
  
  ${({ isLoading }) =>
    isLoading &&
    css`
      & > *:not(${LoadingOverlay}) {
        filter: blur(4px);
      }
    `}
`;