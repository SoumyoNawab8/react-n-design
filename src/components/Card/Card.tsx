import React from 'react';
import { 
  StyledCard, CardCover, CardHeader, CardBody, CardFooter, LoadingOverlay, Spinner 
} from './Card.styles';

export interface CardProps extends React.HTMLAttributes<HTMLDivElement> {
  children: React.ReactNode;
  as?: React.ElementType;
  variant?: 'outset' | 'inset';
  padding?: 'none' | 'small' | 'medium' | 'large';
  /**
   * Adds a subtle border for better contrast.
   */
  bordered?: boolean;
  /**
   * Adds hover and active effects. Useful when the whole card is clickable.
   */
  hoverable?: boolean;
  /**
  * If `true`, a loading spinner will be displayed over the card content.
  */
  loading?: boolean;
  /**
   * Content to render at the top of the card, often an image.
   */
  cover?: React.ReactNode;
  /**
   * Content to render in the header section, above the main children.
   */
  header?: React.ReactNode;
  /**
   * Content to render in the footer section, below the main children.
   */
  footer?: React.ReactNode;
}

/**
 * An advanced neomorphic container with support for covers, headers,
 * footers, and loading states.
 */
export const Card = ({
  children,
  as,
  variant = 'outset',
  padding = 'medium',
  bordered = false,
  hoverable = false,
  loading = false,
  cover,
  header,
  footer,
  ...props
}: CardProps) => {
  return (
    <StyledCard
      as={as}
      variant={variant}
      bordered={bordered}
      hoverable={hoverable}
      isLoading={loading}
      {...props}
    >
      {loading && (
        <LoadingOverlay>
          <Spinner />
        </LoadingOverlay>
      )}
      {cover && <CardCover>{cover}</CardCover>}
      <CardBody padding={padding}>
        {header && <CardHeader>{header}</CardHeader>}
        {children}
        {footer && <CardFooter>{footer}</CardFooter>}
      </CardBody>
    </StyledCard>
  );
};