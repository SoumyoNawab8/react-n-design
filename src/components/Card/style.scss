.rnd-card{
    background-color: #fff;
    padding: 5px 12px 12px 12px;
    margin: 8px;
    border-radius: 6px;
    box-shadow: 1px 2px 15px 0px #8080804d;
    .rnd-card-header{
        padding: 3px;
        display: flex;
        align-items: center;
        justify-content: space-between;
        span{
            font-weight: 600;
        }
        // .rnd-card-close{
        //     height: 15px;
        //     cursor: pointer;
        // }
    }
    .rnd-card-body{
        padding:.5rem 0px;
        overflow: hidden;
    }
}