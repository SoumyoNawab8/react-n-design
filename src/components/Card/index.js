import React from 'react'
import PropTypes from 'prop-types'
import "./style.scss";

function Card({ children, hideHeader = false, title, headerAction, customFooter = null, footer }) {
  return (
    <div className="rnd-card">
      {/* Card Header */}
      {!hideHeader && <div className="rnd-card-header">
        <span className="rnd-card-title">{title}</span>
        <div className="rnd-card-right">{headerAction}</div>
      </div>}
      {/* card body */}
      <div className='rnd-card-body'>{children}</div>
      {/* card footer */}
      <div className="rnd-card-footer">
        {customFooter ? customFooter :
          <div className="rnd-card-default-footer">
            {footer}
          </div>}
      </div>
    </div>
  )
}

Card.propTypes={
  title:PropTypes.string,
  hideHeader:PropTypes.bool,
  customFooter:PropTypes.node,
  footer:PropTypes.node,
  headerAction:PropTypes.node,
  children:PropTypes.node.isRequired,
}


export default Card;