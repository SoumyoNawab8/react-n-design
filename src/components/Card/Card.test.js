import renderer from 'react-test-renderer';
import Card from './index';

it('changes the class when hovered', () => {
  const component = renderer.create(
    <Card title="Henlo">Oki</Card>,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  // re-rendering
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});