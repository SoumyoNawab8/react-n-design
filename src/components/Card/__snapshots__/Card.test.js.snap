// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changes the class when hovered 1`] = `
<div
  className="rnd-card"
>
  <div
    className="rnd-card-header"
  >
    <span
      className="rnd-card-title"
    >
      Henlo
    </span>
    <div
      className="rnd-card-right"
    />
  </div>
  <div
    className="rnd-card-body"
  >
    Oki
  </div>
  <div
    className="rnd-card-footer"
  >
    <div
      className="rnd-card-default-footer"
    />
  </div>
</div>
`;

exports[`changes the class when hovered 2`] = `
<div
  className="rnd-card"
>
  <div
    className="rnd-card-header"
  >
    <span
      className="rnd-card-title"
    >
      Henlo
    </span>
    <div
      className="rnd-card-right"
    />
  </div>
  <div
    className="rnd-card-body"
  >
    Oki
  </div>
  <div
    className="rnd-card-footer"
  >
    <div
      className="rnd-card-default-footer"
    />
  </div>
</div>
`;
